File: src/main/java/net/snipsniper/ImageManager.java
package net.snipsniper;

import java.awt.*;
import java.awt.image.BufferedImage;
import java.io.IOException;
import java.net.URL;
import java.util.HashMap;
import javax.imageio.ImageIO;
import javax.swing.*;

import org.capturecoop.cclogger.CCLogger;
import org.json.JSONArray;
import net.snipsniper.config.ConfigHelper;
import net.snipsniper.utils.FileUtils;
import org.capturecoop.cclogger.CCLogLevel;

public class ImageManager {

	private static final HashMap<String, BufferedImage> images = new HashMap<>();
	private static final HashMap<String, Image> animatedImages = new HashMap<>();
	private static String[] filenameList;

	private ImageManager() { }

	public static void loadResources() {
		CCLogger.log("Loading images...");
		try {
			JSONArray list = new JSONArray(FileUtils.loadFileFromJar("img.json"));
			for(int i = 0; i < list.length(); i++) {
				if(!list.getString(i).endsWith(".gif")) {
					URL url = SnipSniper.class.getResource("/net/snipsniper/resources/img/" + list.getString(i));
					if (url != null) {
						images.put(list.getString(i), ImageIO.read(url));
					} else {
						CCLogger.log("Could not load image %c. This should not happen. Exiting...", CCLogLevel.ERROR, list.getString(i));
						SnipSniper.exit(false);
					}
				} else {
					URL url = ImageManager.class.getResource("/net/snipsniper/resources/img/" + list.getString(i));
					if(url != null) {
						Image img = new ImageIcon(url).getImage();
						animatedImages.put(list.getString(i), img);
					} else {
						CCLogger.log("Could not load image %c. This should not happen. Exiting...", CCLogLevel.ERROR, list.getString(i));
						SnipSniper.exit(false);
					}
				}
			}
		} catch (IOException ioException) {
			CCLogger.log("Could not load resources. Message:", CCLogLevel.ERROR);
			CCLogger.logStacktrace(ioException, CCLogLevel.ERROR);
		}
		CCLogger.log("Done!");
	}

	public static String[] getListAsString() {
		if(filenameList == null) {
			JSONArray list = new JSONArray(FileUtils.loadFileFromJar("img.json"));
			filenameList = new String[list.length()];
			for (int i = 0; i < list.length(); i++) {
				filenameList[i] = list.getString(i);
			}
		}
		return filenameList;
	}

	public static BufferedImage getImage(String path) {
		if(!images.containsKey(path)) {
			CCLogger.log("Could not find image under path " + path + "!", CCLogLevel.ERROR);
			return images.get("missing.png");
		}
		return images.get(path);
	}

	public static Image getAnimatedImage(String path) {
		if(!animatedImages.containsKey(path)) {
			CCLogger.log("Could not find image under path " + path + "!", CCLogLevel.ERROR);
			return images.get("missing.png");
		}
		return animatedImages.get(path);
	}

	public static boolean hasImage(String path) {
		if(animatedImages.containsKey(path))
			return true;
		else return images.containsKey(path);
	}

	public static BufferedImage getCodePreview() {
		switch(SnipSniper.getConfig().getString(ConfigHelper.MAIN.theme)) {
			case "light": return ImageManager.getImage("preview/code_light.png");
			case "dark": return ImageManager.getImage("preview/code_dark.png");
		}
		return null;
	}
}

File: src/main/kotlin/net/snipsniper/ImageManager.kt
package net.snipsniper

import net.snipsniper.config.ConfigHelper
import net.snipsniper.utils.FileUtils
import net.snipsniper.utils.ImageUtils
import org.capturecoop.legiblelogger.LegibleLogger
import org.json.JSONArray
import java.awt.Color
import java.awt.Image
import java.awt.image.BufferedImage
import java.net.URL
import javax.imageio.ImageIO
import javax.swing.ImageIcon

object ImageManager {
    private val images = HashMap<String, BufferedImage>()
    private val animatedImages = HashMap<String, Image>()
    lateinit var filenameList: Array<String>
        private set
    private val missingImgColor = Color(255, 1, 254)
    private val missingImg = getMissingImg()

    fun loadResources() {
        LegibleLogger.info("Loading images...")
        val list = JSONArray(FileUtils.loadFileFromJar("img.json"))
        val filenameListTemp = ArrayList<String>()
        for(i in 0 until list.length()) {
            val str = list.getString(i)
            val url = SnipSniper::class.java.getResource("/net/snipsniper/resources/img/$str") ?: LegibleLogger.warn("Could not load image $str!")
            if(url !is URL) continue

            filenameListTemp.add(str)
            when(str.endsWith(".gif")) {
                true -> animatedImages[str] = ImageIcon(url).image
                false -> images[str] = ImageIO.read(url)
            }
        }
        filenameList = filenameListTemp.toTypedArray()
        LegibleLogger.info("Done!")
    }

    fun getImage(path: String): BufferedImage = images[path] ?: missingImg.also { LegibleLogger.warn("Could not find image under path $path") }
    fun getAnimatedImage(path: String): Image = animatedImages[path] ?: missingImg.also { LegibleLogger.warn("Could not find image under path $path") }
    fun hasImage(path: String): Boolean = animatedImages.containsKey(path) || images.containsKey(path)

    fun getCodePreview() = when(SnipSniper.config.getString(ConfigHelper.MAIN.theme)) {
        "dark" -> getImage("preview/code_dark.png")
        "light" -> getImage("preview/code_light.png")
        else -> missingImg
    }

    private fun getMissingImg(size: Int = 16) = ImageUtils.newBufferedImage(size, size, BufferedImage.TYPE_INT_RGB) {
        var printBlack = true
        for(y in 0 until size) {
            for(x in 0 until size) {
                printBlack = !printBlack
                it.color = if(printBlack) Color.BLACK else missingImgColor
                it.drawLine(x, y, x, y)
            }
            printBlack = !printBlack
        }
    }
}
